File: helpers/read-lines.ts
// @ts-types="npm:@types/node"
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

const DIRNAME = new URL(".", import.meta.url).pathname;
const INPUTS_DIRNAME = "inputs";

export type InputType = "example" | "actual";

export const readLines = (
  day: string,
  year: string,
  type: InputType
): string[] => {
  const filePath = path.resolve(
    DIRNAME,
    "..",
    INPUTS_DIRNAME,
    year,
    `day${day}-${type}.txt`
  );
  const file = fs.readFileSync(filePath, { encoding: "utf-8" });
  const splitted = file.split(os.EOL);
  const t1 = performance.now();
  return splitted;
};

File: helpers/read_lines.go
package helpers

import (
	"bufio"
	"os"
	"path"
)

type InputType string

const (
	Example InputType = "example"
	Actual  InputType = "actual"
)

func ReadLines(day string, year string, inputType InputType) ([]string, error) {
	// TODO(cemreyavuz): construct the file path better
	filename := "day" + day + "-" + string(inputType) + ".txt"

	filepath := path.Join("inputs", year, filename)

	file, err := os.Open(filepath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return lines, nil
}
